Add REST APIs for get, set, update RestConfig

mgao0 opened the pull request:
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

#848 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Added PUT to create, POST to upate, and GET to get the RestConfig for each cluster under CONFIGS.
Tested by unit tests.

### Tests

- [x] The following tests are written for this issue:
testSetRestConfig, testUpdateRestConfig

testCreateClusterRESTConfig, testUpdateClusterRESTConfig, testGetClusterRESTConfig

- [x] The following is the result of the "mvn test" command on the appropriate module:

Tests run: 102, Failures: 0, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  34.405 s
[INFO] Finished at: 2020-03-03T15:48:27-08:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [x] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

zhangmeng916 commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,89 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+    return OK();
+  }
+
+  @GET
```
We may also need a "DELETE" API. You can double check.

mgao0 replied:
I added the "DELETE" API, please help me review again. Thank you.

zhangmeng916 commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,89 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+      return notFound(ex.getMessage());
+    } catch (Exception ex) {
+      _logger.error(
+          "Failed to create rest config, cluster " + clusterId + " new config: " + content
```
try "code reformat" to see whether it'll look better.

mgao0 replied:
I think it's because I used the old style sheet, will update to the new style sheet. Thanks.

pkuwm commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,89 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+    try {
+      record = toZNRecord(content);
+    } catch (IOException e) {
+      _logger.error("Failed to deserialize user's input " + content + ", Exception: " + e);
```
I suggest using `{}` to make advantage of sl4j logging.

mgao0 replied:
I searched for sl4j logging, but couldn't find materials for what you referred to as using{}, could you please provide an example?

zhangmeng916 replied:
I think he meant something like this:
 logger.info("Initialized {} pipeline", Pipeline.Type.TASK.name());

mgao0 replied:
I see. Thank you for explanation @zhangmeng916. @pkuwm Here I'm following the format that is used by other methods in this class. But the way you suggested does provide a clearer and easier way to form the message. Later we can have a separate PR to change all the log messages in Helix to adopt the sl4j {} function.

jiajunwang commented on helix-core/src/main/java/org/apache/helix/ConfigAccessor.java:
```diff
@@ -605,6 +606,45 @@ public RESTConfig getRESTConfig(String clusterName) {
+    String zkPath = scope.getZkPath();
+
+    // Create "/{clusterId}/CONFIGS/REST" if it does not exist
+    String parentPath = HelixUtil.getZkParentPath(zkPath);
```
I see "HelixUtil.getZkParentPath(zkPath)" twice. I guess you are trying to use parentPath.

mgao0 replied:
You are right. I'm trying to use the parentPath. I'm trying to resolve the problem that when we want to update the rest config in the path "/{clusterId}/CONFIGS/REST{clusterId}", the parent path "/{clusterId}/CONFIGS/REST" doesn't exist yet, so I would create the parentPath first.

jiajunwang replied:
What I meant is that line 637 could be:
_zkClient.exists(parentPath)

jiajunwang commented on helix-core/src/main/java/org/apache/helix/ConfigAccessor.java:
```diff
@@ -605,6 +606,45 @@ public RESTConfig getRESTConfig(String clusterName) {
+      ZKUtil.createOrMerge(_zkClient, parentPath, new ZNRecord(parentPath), true, true);
+    }
+
+    if (overwrite) {
```
Actually, if it is "overwrite", there is no need to check existence. Just call createOrReplace().
If it is not "overwrite", there is no need to check existence either, just call createOrUpdate().

mgao0 replied:
Please see the comment above, I checked the existence of parent path, not current path.

jiajunwang replied:
I see.

jiajunwang commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,89 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+    try {
+      configAccessor.setRESTConfig(clusterId, config);
+    } catch (HelixException ex) {
+      return notFound(ex.getMessage());
```
Could it because of write failure? I'm not sure if the accessor could throw any HelixException as well?
I know it probably lists details in the message, but the response type would be notfound in any case, which is not good.

mgao0 replied:
I checked, all the exceptions are caught in ZKUtil.createOrMerge/Replace/Update, and only log the errors, including the write failure or other types of errors. Only one case would throw HelixException which is ZKUtil.isClusterSetup(clusterName, _zkClient) returns false, so I used the notFound response type.  Please let me know if you still think it is an issue and we should fix it some other way. Thanks.

jiajunwang replied:
It's better to have a more precise exception for notFound error. Or, if we use HelixException, we should use a more generic error accordingly.
Since this is a common issue, I'm fine if you add a TODO here. We can address this later.

dasahcc commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,107 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+
+  @POST
+  @Path("{clusterId}/restconfig")
+  public Response updateClusterRESTConfig(@PathParam("clusterId") String clusterId,
```
Can you follow the style of ClusterConfig? We have entry update/delete operation with URL param.

mgao0 replied:
I have updated the PR. Can you review again? Thanks.

dasahcc replied:
I think you have missunderstanding about delete the Config and delete entry of the Config.

dasahcc commented on helix-core/src/main/java/org/apache/helix/ConfigAccessor.java:
```diff
@@ -605,6 +606,62 @@ public RESTConfig getRESTConfig(String clusterName) {
+    updateRESTConfig(clusterName, restConfig, false);
+  }
+
+  private void updateRESTConfig(String clusterName, RESTConfig restConfig, boolean overwrite) {
```
Is this necessary? We should have update function with scope arugment,right?

mgao0 replied:
I've adopted the pattern for ClusterConfig here. The overwrite is to be called to distinguish set and update operations. The RESTConfig instance passed in gives us the information to be used to set/update the rest config, and the scope can be derived from information in RESTConfig in the implementation.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,125 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+
+    if (!record.getId().equals(clusterId)) {
+      return badRequest("ID does not match the cluster name in input!");
+    }
```
Could we double-check if the ID for the restConfig is supposed to match up with the cluster name?

mgao0 replied:
This check is not imposed on rest config id, this is to check the if the id of ZNRecord user uses to update the rest config matches the cluster name. I'm not sure what our users are doing right now, but this could prevent accidentally update the wrong cluster.

mgao0 replied:
Since it is a not necessary check. I removed it so it is easier for users to use.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,125 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
 
+  @PUT
+  @Path("{clusterId}/restconfig")
+  public Response createClusterRESTConfig(@PathParam("clusterId") String clusterId,
```
Nit: is there a reason we don't just say createRestConfig? Cluster seems a little redundant?

mgao0 replied:
I've changed the method names.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,125 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+  @Path("{clusterId}/restconfig")
+  public Response getClusterRESTConfig(@PathParam("clusterId") String clusterId) {
+    ConfigAccessor accessor = getConfigAccessor();
+    RESTConfig config = null;
```
Initializing to null is redundant.

mgao0 replied:
I initialize the config to null mainly to check later if the accessor got the config successfully. If there is an Exception, we catch these exceptions, so by comparing if the config is null we know if there is error in fetching the information.

narendly replied:
Please read https://docs.oracle.com/javase/specs/jls/se7/jls7.pdf

Object references not initialized default to null. So initializing to null is redundant.

mgao0 replied:
I tried it, the code doesn't compile after removing "=null". I agree that initialization to null is functional redundant, but it benefits the style check and readability of the code, so I would keep it here.

dasahcc commented on helix-core/src/main/java/org/apache/helix/ConfigAccessor.java:
```diff
@@ -605,6 +606,62 @@ public RESTConfig getRESTConfig(String clusterName) {
+    updateRESTConfig(clusterName, restConfig, false);
+  }
+
+  private void updateRESTConfig(String clusterName, RESTConfig restConfig, boolean overwrite) {
```
There are lots of common code for updateRESTConfig, updateClusterConfig and so on. I am fine with this PR. But can you have another PR to abstract the logic to a private method to reduce the redundant code?

mgao0 replied:
Ok. I'll create another PR to abstract the logic.

narendly commented:
Please fix/improve the code where change is possible. The fact that existing code is doing it in a certain suboptimal way doesn't justify us making the same mistake. I am fine, however, with creating a separate issue for refactoring work that will involve changes in existing logic though.

pkuwm commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:
```diff
@@ -511,6 +512,121 @@ public Response removeClusterStateModelDefinition(@PathParam("clusterId") String
+    try {
+      accessor.deleteRESTConfig(clusterId);
+    } catch (HelixException ex) {
+      _logger.info("Failed to delete rest config for cluster " + clusterId
```
Failed to delete but finally returning OK response? Would 404 be more appropriate?

mgao0 replied:
I was considering returns OK would be fine since what you want to delete is already not there. But on second thought, return NOT_FOUND probably is more intuitive for the users. Changing it to NOT_FOUND.

pkuwm commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestClusterAccessor.java:
```diff
@@ -576,6 +577,70 @@ public void testEnableWagedRebalanceForAllResources() {
+    Assert.assertEquals(restConfigZK, new RESTConfig(cluster),
+        "rest config from response: " + new RESTConfig(cluster) + " vs rest config actually: "
+            + restConfigZK);
+    System.out.println("End test :" + TestHelper.getTestMethodName());
```
Is the test already ended?

mgao0 replied:
Added one more test after the first draft. Moving it down.

mgao0 commented:
> Please fix/improve the code where change is possible. The fact that existing code is doing it in a certain suboptimal way doesn't justify us making the same mistake. I am fine, however, with creating a separate issue for refactoring work that will involve changes in existing logic though.

Rolling out functional service for our customers at the earliest to unblock them is our top priority. We proved the current way works fine, and the new way although is clearer in code, it bears the risk of breaking things for our customers. Therefore, I considered it sub-priority for us. Also, refactoring for one thing all together is easier for us to track and debug if anything goes wrong in the future.

narendly commented on helix-core/src/main/java/org/apache/helix/ConfigAccessor.java:
```diff
@@ -604,6 +605,61 @@ public RESTConfig getRESTConfig(String clusterName) {
     return new RESTConfig(record);
   }
 
+  /**
```
If there is a change to config accessor, should we add tests in TestConfigAccessor as well?

mgao0 replied:
There are tests for rest config covering set and update in TestConfigAccessor. I just added a few more cases for delete.

narendly commented on helix-core/src/test/java/org/apache/helix/TestConfigAccessor.java:
```diff
@@ -204,4 +207,56 @@ public void testSetNonexistentParticipantConfig() throws Exception {
+    Assert.assertEquals(restConfig, configAccessor.getRESTConfig(clusterName));
+  }
+
+  @Test (expectedExceptions = HelixException.class)
```
Could you mark it clearly where the exception should be expected? Because it's not often easy to tell, I prefer the style of 

```
try {
//your logic
Assert.fail("Exception expected!");
catch (exception) {
// Exception expected..
}
```

mgao0 commented:
This PR is ready to be merged, approved by @narendly .
Final commit message: 
Title: Add REST APIs for get, set, update RestConfig
Body: Added endpoints PUT to create, POST to update, and GET to get, DELETE to delete the RestConfig for clusters in ClusterAccessor. Added corresponding methods in ConfigAccessor to be used by ClusterAccessor.
